FF7: Add native support for japanese text rendering

CosmosXIII opened the pull request:
## Summary

Adds an option to enable japanese text rendering for the japanese version of the game.
Right now field text only works if the jp flevel is re-exported using Makou Reactor.

Related #39 

### Motivation

To end the pain of the japanese nation. 

### ACKs

- [x] I have updated the [Changelog.md](https://github.com/julianxhokaxhiu/FFNx/blob/master/Changelog.md) file
- [x] I did test my code on FF7
- [ ] I did test my code on FF8

CosmosXIII committed:
FF7: Added option for japanese text rendering

CosmosXIII committed:
FF7: Added option for japanese text rendering

julianxhokaxhiu commented on Changelog.md:
```diff
@@ -13,6 +13,7 @@
 ## FF7
 
 - Core: Add missing Cloud models in the model to eye mapping list
+- Rendering: Added option for japanese text rendering ( https://github.com/julianxhokaxhiu/FFNx/pull/737 )
```
Can you please move this into the Next section?

CosmosXIII replied:
Done, somehow I always put it in the wrong place lol

julianxhokaxhiu committed:
Gardening: Update Changelog and install docs

julianxhokaxhiu committed:
Use ff7_japanese_edition as flag

julianxhokaxhiu commented:
Tested the PR on top of the native Japanese eStore release and first of all, well done :) I'm really impressed by this achievement!
I mainly noticed some bugs while using it, at least visually, I'll try to list them here so we can have a look together on how to fix it in case.

1. Save slots are not aligned with the hand, here an example:
JA:
![image](https://github.com/user-attachments/assets/6151ee48-8659-4635-a481-44247421c52b)
EN:
![image](https://github.com/user-attachments/assets/f318b468-e244-43ca-b359-2dccf14d004a)

2. Location text seems to be wrong/broken(?) I can see this in both save menu as well as in-game menu:
![image](https://github.com/user-attachments/assets/9acff62d-9cf3-4167-b8c8-bac865468f68)
![image](https://github.com/user-attachments/assets/539c3ae6-89d6-4725-8a93-6f9ae086962a)

3. Loading text seems also to be broken(?)
![image](https://github.com/user-attachments/assets/3a50b1c7-b1ff-46ea-9c58-2fba0475629a)

I'm not sure if the context of those makes sense as I don't know the language, so any help on your side to check that is appreciated :)

Thanks a lot and looking forward to this really promising PR!

P.S.: I changed a bit the logic on how we enable this layer. Your original flag has been removed in favour of `ff7_japanese_edition` as it was already there since the beginning of the code, and we auto-enable it when we detected the japanese version of the game. However I left it also as a flag in the toml in case we might need to enable this manually because the auto-detection fails. Hope this is fine for you, let me know in case. Finally, I updated a bit the changelog message to fix the overall mission of this PR, let me know what you think.

CosmosXIII commented:
Thanks for testing it.

> 1. Save slots are not aligned with the hand

For the iro I was replacing the exe text with hext so I just added a few spaces to fix this.
By the way the english exe text also has 3 spaces like this: "Save 1   "
I'm not sure why but the japanese one did not have spaces so my guess is they hardcoded the spacing somewhere in code...

> 2. Location text seems to be wrong/broken(?)

I think this is because of the broken flevel. Once I fixed that with Makou the location also worked fine for me.

> 3. Loading text seems also to be broken(?)

The text here actually is correct but it's going outside the box.
So I think the fix here is to resize the box so that it fits the text but no idea where this is done.

julianxhokaxhiu committed:
Enable the flag only if the exe is ff7_ja

If not, inherit what was set via the config

CosmosXIII committed:
Fixed loading text box size

CosmosXIII committed:
Improved text box auto-resize to take into account lenght of character names

julianxhokaxhiu committed:
Update Changelog.md

canesalato commented:
Thank you so much for this work Cosmos!
Will this require a repurchase of the game? Japanese version? Or will the steam version be able to be converted to support Japanese language, like with FF9 and memoria?

CosmosXIII commented:
> Thank you so much for this work Cosmos! Will this require a repurchase of the game Japanese version or will the steam version be able to be converted to support Japanese language, like with FF9 and Memoria?

Unlike FF9 the JP data is not included  in the Steam release but it's technically possible to make it work with the Steam version if you provide the JP data in an IRO mod. Not sure yet if we it will go in that direction though.

zaphod77 commented:
the android version seems to have broken flevel files too.  even the english one. at least the version I got.

the windows store version does not have broken flevel files.

How exactly did you do the makou reactor fix? i tried editig a text so makour rector would save, and it didn't fix a thing.

julianxhokaxhiu commented:
> How exactly did you do the makou reactor fix? i tried editig a text so makour rector would save, and it didn't fix a thing.

Please do not hijack this PR with offtopic questions. If you need support or help, use Discord support channels. Thank you

zaphod77 commented:
the new ffnx resolves all broken lgp file issues including the one that's messing with this. the jp e-store release jfleve.lgp should read fine now, like the android version.

julianxhokaxhiu commented:
@zaphod77 that was never the issue blocking this PR. Read above what was pending for.

CosmosXIII commented:
> @zaphod77 that was never the issue blocking this PR. Read above what was pending for.

I think zap is referring about what I wrote in the PR summary about field text not working with unmodified flevel unless it’s re-exported with MR. I’ll check later to see if it works.

canesalato commented:
I have been testing this PR with the mobile version of the game.

To start with, I can confirm that, before the latest changes @zaphod77, field text was completely broken. It used to display actual Japanese characters, but making meaningless words with it. I am very pleased to report that everything seems to be working well now. The game plays just fine in Japanese. I checked all the main menus, tried different resolutions, 16:9 vs 4:9, loaded different save files and, so far, it seems to be working very well. Also I did not see major misalignement or issues (with 2 exceptions see points 1 and 4).

Concerning the points (1-3) raised by @julianxhokaxhiu:
1. This still happens, and I found it in the field as well (see attachment). Basically, everytime there is a selection to make, the cursor is slightly misaligned. It seems indeed to be related to some hardcoded spacing value.
![screenshot (16)](https://github.com/user-attachments/assets/48cbb794-f0c2-4b5c-b5ae-ff5bf13375bb)
![screenshot (18)](https://github.com/user-attachments/assets/e05b7778-2b59-47c6-96ad-29d3228139ab)
2.  Fixed for me after @zaphod77 's latest commit! (see attachment) 
![Screenshot 2024-12-05 122715](https://github.com/user-attachments/assets/e67116c7-a614-4771-a582-a3e27973fa70)
![Screenshot 2024-12-05 122615](https://github.com/user-attachments/assets/7e95eb16-3ea6-4c94-8d54-ab551b835428)
3.  Loading screen also looks fixed now!

Something I found, not previously reported:
4. An issue with text when displaying input keys: eg. playing the piano at Tifa's house (see attachment):
![screenshot (17)](https://github.com/user-attachments/assets/c8055a25-d556-43bb-84bd-bb4cb58b784f)
![screenshot (19)](https://github.com/user-attachments/assets/bc69c540-7cf6-4c77-a279-971472a4287f)
5. Finally, I have noticed that Japanese text seems to work well even with the japanese edition flag set to "false" in the configuration file. However, setting it to "true", seems to break text when loading the normal english executable.

canesalato commented:
A couple of additional instances where things go weird.

1. Gold Saucer, Chocobo racing, Chocobo selection dialog
![PXL_20241214_052139917](https://github.com/user-attachments/assets/211029b2-f80b-4acd-a524-29d9571436f5)
2. Gold Saucer, Mog video game, at game over
![PXL_20241214_052643520](https://github.com/user-attachments/assets/73b24f50-f4b4-4278-a92e-75980067171f)

canesalato commented:
Some updates that perhaps can be interesting. In my previous post I wrote that there seem to be an issue with displaying input keys (like, tifa's home piano). 

I would like to partially revise that statement. 
It does not seem to be specifically input keys, it applies to everything that has color. At first I missed this, because the english version seem to be missing much of the text color of the japanese version. You can compare the Vanilla Japanese and English PC versions
![PXL_20250204_142756835](https://github.com/user-attachments/assets/cbb9c4b1-a3de-4a6f-9374-62acbd112607)
![PXL_20250204_133340501](https://github.com/user-attachments/assets/e28f78f4-cab2-48b9-b0ec-4c19115fce9b)

And this is how it appears in the Japanese FFNx from this PI:
![screenshot (8)](https://github.com/user-attachments/assets/f94f0ad1-d8d3-4b89-a7f4-f041fc90ba95)
Notice that the font is white and the セ from セーブ is missing. This seem to be a constant with all colored text. Could the problem be that some colored font textures are missing?

Here is another example:
![![screenshot (9)](https://github.com/user-attachments/assets/cb8a1271-b6d5-4bae-aa51-37a09940b3e5) PXL_20250204_133031230.jpg…]()

julianxhokaxhiu committed:
FF7: Fix field_78 field access

canesalato commented:
The 2026 Steam re-release includes Japanese language  support. Is their af3dn.p different from the JP e-store?


________________________________
From: zaphod77 ***@***.***>
Sent: Saturday, March 14, 2026 09:17
To: julianxhokaxhiu/FFNx ***@***.***>
Cc: canesalato ***@***.***>; Comment ***@***.***>
Subject: Re: [julianxhokaxhiu/FFNx] FF7: Add native support for japanese text rendering (PR #737)

[https://avatars.githubusercontent.com/u/2146277?s=20&v=4]zaphod77 left a comment (julianxhokaxhiu/FFNx#737)<https://github.com/julianxhokaxhiu/FFNx/pull/737#issuecomment-4058869209>

more fun.

image.png (view on web)<https://github.com/user-attachments/assets/6d6d4090-346a-44f0-bed5-f3421a2eaeb4>

Unfortunately, will probably need to reverse engineer the af3dn.p from jp estore release to work out what the @$@#% they did for the naming screen, and to solve the rest of the issues. Even if i knew how to decompile a driver, i don't have the JP e-store release, and it can't be bought now.

—
Reply to this email directly, view it on GitHub<https://github.com/julianxhokaxhiu/FFNx/pull/737#issuecomment-4058869209>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASWUKDODC3JB4IBWCRZAIUD4QSQK5AVCNFSM6AAAAACUZ4TOLGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DANJYHA3DSMRQHE>.
You are receiving this because you commented.Message ID: ***@***.***>

zaphod77 commented:
I'm 100% okay with you taking the japanese_text.cpp from my file, modifying it, and improving this pull request.  Only reason it's not here is that I was unable to build your fork when i tried to fork it, ok?

TurquoiseHammer commented:
Mmyellow. Going about testing SAC's JP version and I've noticed the JP in non-field windows is a smaller point across the board. Here are a couple comparisons.

Main menu launching through Steam:
<img width="1261" height="699" alt="image" src="https://github.com/user-attachments/assets/29846f8b-a85f-401b-9a8f-b18aaacfe2e6" />
Main menu launching through 7H:
<img width="969" height="710" alt="image" src="https://github.com/user-attachments/assets/04f375c0-4ed6-4fba-bd1d-04da08a2d45f" />

Steam shop window:
<img width="981" height="194" alt="image" src="https://github.com/user-attachments/assets/f8b6523e-7395-430c-a0ac-fd821fc9e78e" />
7H shop window:
<img width="923" height="290" alt="image" src="https://github.com/user-attachments/assets/1ebe0bd9-f93a-49c7-8e8c-a5ce03a1cf95" />

I can test more locations if you'd like, but it seems to hold true for anything off the field.

zaphod77 commented:
I have known this for a while. the game uses two sizes, and there's no way to work out which one is supposed to be used with the existing externals.  what seems to matter is the source of the text.

Some texts are small and some texts are big. color and what menu screen it is alone is not enough to guess the proper size.

TurquoiseHammer commented:
Hmm so are you saying it would just involve a lot of trial and error blindly setting font sizes to see if they increase the right thing?

zaphod77 commented:
The issue is the function we are hooking isn't sent the information we need to make a decision on the text size.

https://github.com/julianxhokaxhiu/FFNx/blob/7717ef454936127b3b83b004b1103e31dd903e4b/src/ff7/japanese_text.cpp#L1114

This function only let's us access the character string, the position, the color, the z order, and the memory location the text is being read from.  *That's not enough information to determine the size*, because the text is **copied into a buffer first before it's sent to this function**, so we have no idea what we are actually trying to print.  We need to intercept the copying of the text to the buffer, and change the scale factor.  We know field dialogs are always big, so that works fine, but the other values have mixes of big and small.  I spent ages trying to be clever, and simply couldn't get anywhere, and had to make everything but some stuff we cold safely determine is big to small.  The original square code also applies alignment changes, and itself isn't even perfect.

I had hoped it was a L vs H difference (and that would have made some SENSE) but implementing checking that did absolutely nothing useful.

TurquoiseHammer commented:
Okay sounds messy. I'm not a coder, or at least not a skilled one, so forgive me if these are silly questions:
-This function handles all text drawing, right? And I assume from the parameters passed in, you can determine if it's field text, in which case you make it "big."
-So if it's menu text, would a potential stopgap be to see if it matches any of the strings we're expecting in certain menus? I guess this could work for some things like Items, Materia, Save, etc., but would get pretty out of hand if you want to check every single item name, every map name, etc.

zaphod77 commented:
NO i can't.  

the field dialogue one doesn't call that, so that's why it's big like it's supposed to be.  But most other battle menu and other menu stuff does call that function.   and given the choice between making everything big vs making everything small, i had to pick small so stuff will fit.

It not the string that determines big/small, square hooks the menu modules, and changes the size and alignment based on what the game is trying to print. whether it's a player name, status text, item name, etc.

For example sort popup and the three terxts at the upper left of use item menu are small, but the rest is big.

But the character stats menu is a MESS, and a mix of texts of different colors and sizes.

TurquoiseHammer commented:
Thanks, I'm getting a better understanding now. I appreciate you explaining it all. I guess what confused me is you said "This function only let's us access the character string, the position, the color, the z order, and the memory location the text is being read from." Is this information not enough to either a) change the scale factor based on the memory location or b) check if the string matches one of the small strings in the menu, such as つぎのレベルまであと?

zaphod77 commented:
there are way too many strings to do that. it's the class of text, and the module that's printing it that matters.

For example, Item names in the item menu are always big.  

And changing the scale factor based on the memory location doesn't work, as i said, because the text is coped from its original source to a temporary buffer to use with the function. so i can't find out where the text was copied from.

Basically exe texts have to be assigned a size attribute that they don't actually have.  and while battle win and us font DO have two different versions, the jp font does not.

TurquoiseHammer commented:
Ah, okay I understand the memory location thing now w/r/t the buffer.
Last thought: what if you did it the other way around—check for things that should be small and make the rest big. It seems like there are fewer of those, and they're never variable as far as I know (character names, items, descriptions, etc). (And I'm aware that this is awfully hacky. Was just thinking as a temporary measure.)

zaphod77 commented:
It's very annoying.  You need to check all menus in the jp pc version to see how crazy some of those screens get with mixed text.

TurquoiseHammer commented:
I forgot that the spell names in the bottom left panel of the materia submenu are small text. Yeah that would be hundreds of strings to manually check against... Well thanks all the same. Let me know if there's anything I can do to help, even if it's something brute-forcey.
I've got a couple other questions about text-related stuff, hopefully more straightforward ones with simple fixes.

First, on certain screens like this one, FFNx doesn't pad the text the same way the official port does.
FFNX: <img width="950" height="697" alt="1" src="https://github.com/user-attachments/assets/8269f810-37b7-44c4-8fd0-1b40ac7f8ee9" />
Official port: 
<img width="991" height="692" alt="2" src="https://github.com/user-attachments/assets/50bcbccd-f46c-40aa-a642-4b289357d715" />
Of course, the ideal would be perfectly centered like the PSX, but that will have to come if and when we can implement a wider font
<img width="1056" height="743" alt="3" src="https://github.com/user-attachments/assets/d8c033ef-afac-4727-ac7d-9f762269794e" />

These lines are just manually spaced in the flevel, so I wonder if something is reducing that whitespace here.
<img width="219" height="131" alt="image" src="https://github.com/user-attachments/assets/6982c33f-9c46-4382-8023-ef71bdc949b2" />


Next, and possibly the same issue with whitespace getting eaten, lines in FFNx JP aren't flush with the inside of opening quotation marks like they're supposed to be.
FFNX (ignore that heart symbol, that's a different problem): 
<img width="945" height="461" alt="image" src="https://github.com/user-attachments/assets/dcac4ca1-4e5d-4cd1-8f3f-1dded81f817f" />
Official port:
<img width="862" height="443" alt="image" src="https://github.com/user-attachments/assets/0ef3848d-3c78-4f52-b04f-4dfe8b69929a" />

zaphod77 commented:
You've actually found a bug somehow. this was only supposed to put in the heart if it was on the first ja_font.

TurquoiseHammer commented:
Oh haha, I thought it was SAC related.

zaphod77 commented:
made a pull request to fix it

TurquoiseHammer commented:
Ok I looked into it a little more and the issue is that in FFNx JP, the space (　) character is rendered too narrow. This throws off the spacing on dialogue choices and other flashback dialogue/internal monologue stuff that uses that character to pad text. Fixing that should fix all these problems at once.

This is how a standard four-space-padded dialogue option should look:
<img width="822" height="607" alt="image" src="https://github.com/user-attachments/assets/607de5bf-87d0-4f9f-9cc5-ba7fbed7354e" />
And this is how it looks in FFNx right now. Not a big issue, but sometimes event planners only use three spaces, and this causes the cursor to overlap the text on FFNx.
<img width="934" height="691" alt="image" src="https://github.com/user-attachments/assets/1de257cc-b5f9-4034-bc50-76fad207d687" />

zaphod77 commented:
that should be simple as going to the width table that starts at line 320, and editing the number for the space.

zaphod77 commented:
Okay, setting space at 31 width is too big for sure. 24 seems closer

also there's not enough padding overall.

You can see that the text strings look longer in the official port in the field dialogs.

TurquoiseHammer commented:
That's because dialogue window autosizing is turned off in the version I'm testing, so the box is wider. The text lines up perfectly otherwise, aside from the tiny bit of missing space at the beginning of the line.

zaphod77 commented:
No, that's not it.  i'm talking about the actual text. it seems the max width is supposed to be 32, not 31.

TurquoiseHammer commented:
Oh, do you mean whatever is happening here?
<img width="1221" height="424" alt="image" src="https://github.com/user-attachments/assets/8d419879-864a-468d-bca4-71672f3c31c2" />
(Longer text is official port)

It looks like the character width on っ and た is every so slightly narrower

zaphod77 commented:
It's obvious in the meeting screenshot.

ter's a little more padding between characters.

the impression i get is that all non zero padding values have to have one added to them to fix this.

TurquoiseHammer commented:
Here's a better comparison of the meeting room text windows:
<img width="1172" height="629" alt="image" src="https://github.com/user-attachments/assets/9a774f4b-6125-442d-9bfa-ce5cd03f2fb0" />

second and third lines, making up for reduced space character:
<img width="1142" height="573" alt="image" src="https://github.com/user-attachments/assets/0b052037-c600-4c0b-a2c3-77222becb2a4" />

So that's just getting offset by one pixel per character?

zaphod77 commented:
looks like. but fixing that with a patch i made throws off the auto resizer.

so i have to patch that too.

but THAT will break the arabic, which expects the current behavior.

TurquoiseHammer commented:
Imo it's not a huge issue to leave it as-is. The auto sizing accounts for it, right? And it makes no difference to legibility. Plus, ideally at some point we will introduce a whole new JP font that works more like the original game's.

zaphod77 commented:
Not doing the offset DOES affect legibility.

And not adjusting the autosize to treat characters as one wider than they are if <32 causes it to not move the right side of the box enough.

This is VERY annoying.

TurquoiseHammer commented:
Okay here is what confuses me though. If it were a padding issue between characters, wouldn't the third line here offset at the same rate as the second? And yet all the ドs line up 
<img width="842" height="388" alt="image" src="https://github.com/user-attachments/assets/2d4b9a02-abfd-48dd-8b09-232ae8df648f" />

zaphod77 commented:
you are right.  it seems square is just doing all sorts of weird stuff and hacky alignment fixes.  sigh...

julianxhokaxhiu commented:
Please move on the conversation on the issue, which I've also left some notes which you might find interesting: https://github.com/julianxhokaxhiu/FFNx/issues/39#issuecomment-5348104829

TurquoiseHammer commented:
Confirming the 値 character no longer results in a heart!
<img width="717" height="327" alt="image" src="https://github.com/user-attachments/assets/670c7203-61b4-4e9a-87cd-97235f222f8c" />
